compiler/rustc_middle/src/ty/closure.rs

```diff
@@ -155,6 +155,54 @@ impl CapturedPlace<'tcx> {
         place_to_string_for_capture(tcx, &self.place)
     }
 
+    /// Returns mangled names of captured upvars. Here are some examples:
+    ///  - `_captured_val__name__field`
+    ///  - `_captured_ref__name__field`
+    ///
+    /// The purpose is to use those names in debuginfo. They should be human-understandable.
+    /// Without the names, the end users may get confused when the debuggers just print some
+    /// pointers in closures or generators.
+    pub fn to_mangled_name(&self, tcx: TyCtxt<'tcx>) -> String {
+        let prefix = match self.info.capture_kind {
+            ty::UpvarCapture::ByValue(_) => "_captured_val__",
+            ty::UpvarCapture::ByRef(_) => "_captured_ref__",
+        };
+
+        let hir_id = match self.place.base {
+            HirPlaceBase::Upvar(upvar_id) => upvar_id.var_path.hir_id,
+            base => bug!("Expected an upvar, found {:?}", base),
+        };
+        let name = tcx.hir().name(hir_id);
+
+        let mut ty = self.place.base_ty;
+        let mut fields = String::new();
+        for proj in self.place.projections.iter() {
+            match proj.kind {
+                HirProjectionKind::Field(idx, variant) => match ty.kind() {
+                    ty::Tuple(_) => fields = format!("{}__{}", fields, idx),
+                    ty::Adt(def, ..) => {
+                        fields = format!(
```

Comment:
Instead of rebuilding the string here, extend it using `write!()`

(you can unwrap the result, String never fails)

Reply:
OK, I'll do it.